File: src/tools/webfetch.ts
// WebFetch tool — fetch a URL and return its readable text content.
//
// Companion to web_search: the model searches, gets result URLs, then fetches
// one for the full page. Zero key / zero config — a plain HTTPS GET.
//
// Like web_search this is read-only and network-only (no filesystem, no
// approval gate). Trade-offs we accept, not solve:
//   - We strip HTML to text with regexes, not a real DOM. Good enough for
//     articles/docs; messy on JS-rendered SPAs (which return little markup).
//   - http:// is upgraded to https:// before the request (mirrors Claude
//     Code) — avoids accidental plaintext fetches and most redirects.
//   - Output is truncated to MAX_CHARS so one fetch can't blow the context.

import type { ToolResult } from "./executor.js";

const UA =
  "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) " +
  "AppleWebKit/537.36 (KHTML, like Gecko) Chrome/124.0.0.0 Safari/537.36";

const CACHE_TTL_MS = 15 * 60 * 1000;
const REQUEST_TIMEOUT_MS = 15_000;
const MAX_CHARS = 50_000;

const cache = new Map<string, { at: number; content: string }>();

/** Decode the HTML entities that actually show up in body text. */
function decodeEntities(s: string): string {
  return s
    .replace(/&#x27;/g, "'")
    .replace(/&#39;/g, "'")
    .replace(/&quot;/g, '"')
    .replace(/&nbsp;/g, " ")
    .replace(/&mdash;/g, "—")
    .replace(/&ndash;/g, "–")
    .replace(/&hellip;/g, "…")
    .replace(/&lt;/g, "<")
    .replace(/&gt;/g, ">")
    .replace(/&amp;/g, "&");
}

/**
 * Convert an HTML document to readable plain text. Exported for testing.
 *
 * Not a parser — a deliberately small pipeline: drop non-content elements,
 * turn block-level tags into newlines and links into "text (url)", strip
 * what's left, then collapse whitespace.
 */
export function htmlToText(html: string): string {
  let s = html;

  // Drop elements whose text content is never page content.
  s = s.replace(
    /<(script|style|noscript|template|svg|head)\b[\s\S]*?<\/\1>/gi,
    " ",
  );
  s = s.replace(/<!--[\s\S]*?-->/g, " ");

  // Preserve link targets: <a href="X">label</a> → "label (X)".
  s = s.replace(
    /<a\b[^>]*\bhref=['"]([^'"]+)['"][^>]*>([\s\S]*?)<\/a>/gi,
    (_m, href: string, label: string) => {
      const text = label.replace(/<[^>]+>/g, "").trim();
      if (!text) return "";
      return href.startsWith("http") ? `${text} (${href})` : text;
    },
  );

  // Block-level boundaries → newlines so structure survives tag stripping.
  // <li> opens its own bullet line, so it's excluded here to avoid a
  // blank line between every list item from a doubled newline.
  s = s.replace(/<\/(p|div|section|article|tr|h[1-6]|blockquote)\s*>/gi, "\n");
  s = s.replace(/<(br|hr)\s*\/?>/gi, "\n");
  s = s.replace(/<li\b[^>]*>/gi, "\n- ");

  // Strip all remaining tags, decode entities.
  s = s.replace(/<[^>]+>/g, "");
  s = decodeEntities(s);

  // Collapse runs of blank lines and trailing whitespace.
  s = s
    .split("\n")
    .map((line) => line.replace(/[ \t\f\v]+/g, " ").trim())
    .join("\n")
    .replace(/\n{3,}/g, "\n\n")
    .trim();

  return s;
}

/** Normalize/validate the URL: require http(s), upgrade http→https. */
function normalizeUrl(raw: string): string | null {
  let u: URL;
  try {
    u = new URL(raw.trim());
  } catch {
    return null;
  }
  if (u.protocol !== "http:" && u.protocol !== "https:") return null;
  if (u.protocol === "http:") u.protocol = "https:";
  return u.toString();
}

/**
 * Fetch a URL and return its content as text. Read-only and network-only —
 * never touches the filesystem, so it needs no workspace and no approval gate.
 */
export async function executeWebFetch(
  args: Record<string, unknown>,
): Promise<ToolResult> {
  const url = normalizeUrl(String(args.url ?? ""));
  if (!url) {
    return {
      content: "Error: a valid http(s) url is required.",
      isError: true,
    };
  }

  const cached = cache.get(url);
  if (cached && Date.now() - cached.at < CACHE_TTL_MS) {
    return { content: cached.content, isError: false };
  }

  const ctrl = new AbortController();
  const timer = setTimeout(() => ctrl.abort(), REQUEST_TIMEOUT_MS);
  let res: Response;
  try {
    res = await fetch(url, {
      headers: { "User-Agent": UA, Accept: "text/html,*/*" },
      redirect: "follow",
      signal: ctrl.signal,
    });
  } catch {
    clearTimeout(timer);
    return {
      content: `Error: failed to fetch ${url} (network error or timeout).`,
      isError: true,
    };
  }
  clearTimeout(timer);

  if (!res.ok) {
    return {
      content: `Error: ${url} returned HTTP ${res.status} ${res.statusText}.`,
      isError: true,
    };
  }

  const ctype = res.headers.get("content-type") ?? "";
  const raw = await res.text();
  // HTML → readable text; JSON/plain text/markdown pass through verbatim.
  const body = /\bhtml\b/i.test(ctype) ? htmlToText(raw) : raw.trim();

  if (!body) {
    return {
      content: `Fetched ${url} but it had no extractable text content.`,
      isError: false,
    };
  }

  const truncated =
    body.length > MAX_CHARS
      ? body.slice(0, MAX_CHARS) +
        `\n\n[truncated — ${body.length - MAX_CHARS} more chars]`
      : body;
  const content = `Content of ${url}:\n\n${truncated}`;

  cache.set(url, { at: Date.now(), content });
  return { content, isError: false };
}

File: deepdive-rs/crates/deepdive-core/src/tools/webfetch.rs
//! WebFetch tool — fetch a URL and return readable text. Faithful port of
//! `src/tools/webfetch.ts`. Read-only + network-only.
//!
//! The TS script/style drop uses a `</\1>` backreference; since the tag set is
//! fixed we expand it into one regex per tag (no backreference needed, so the
//! plain `regex` crate suffices). The `<a>` rewrite uses a closure replacer.

use crate::tools::executor::ToolResult;
use regex::{Captures, Regex};
use std::collections::HashMap;
use std::sync::{LazyLock, Mutex};
use std::time::{Duration, Instant};

const UA: &str = "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/124.0.0.0 Safari/537.36";
const CACHE_TTL: Duration = Duration::from_secs(15 * 60);
const REQUEST_TIMEOUT: Duration = Duration::from_millis(15_000);
const MAX_CHARS: usize = 50_000;
const THIN_TEXT: usize = 200;

static CACHE: LazyLock<Mutex<HashMap<String, (Instant, String)>>> =
    LazyLock::new(|| Mutex::new(HashMap::new()));

const DROP_TAGS: &[&str] = &["script", "style", "noscript", "template", "svg", "head"];
static DROP_RES: LazyLock<Vec<Regex>> = LazyLock::new(|| {
    DROP_TAGS
        .iter()
        .map(|t| Regex::new(&format!(r"(?is)<{t}\b.*?</{t}>")).unwrap())
        .collect()
});
static COMMENT_RE: LazyLock<Regex> = LazyLock::new(|| Regex::new(r"(?s)<!--.*?-->").unwrap());
static LINK_RE: LazyLock<Regex> = LazyLock::new(|| {
    Regex::new(r#"(?is)<a\b[^>]*\bhref=['"]([^'"]+)['"][^>]*>(.*?)</a>"#).unwrap()
});
static TAG_RE: LazyLock<Regex> = LazyLock::new(|| Regex::new(r"<[^>]+>").unwrap());
static BLOCK_CLOSE_RE: LazyLock<Regex> = LazyLock::new(|| {
    Regex::new(r"(?i)</(p|div|section|article|tr|h[1-6]|blockquote)\s*>").unwrap()
});
static BR_RE: LazyLock<Regex> = LazyLock::new(|| Regex::new(r"(?i)<(br|hr)\s*/?>").unwrap());
static LI_RE: LazyLock<Regex> = LazyLock::new(|| Regex::new(r"(?i)<li\b[^>]*>").unwrap());
static INLINE_WS_RE: LazyLock<Regex> = LazyLock::new(|| Regex::new(r"[ \t\x0c\x0b]+").unwrap());
static BLANKS_RE: LazyLock<Regex> = LazyLock::new(|| Regex::new(r"\n{3,}").unwrap());
static BLOCKED_RE: LazyLock<Regex> = LazyLock::new(|| {
    Regex::new(r"(?i)\b(enable javascript|requires javascript|javascript is (disabled|required)|checking your browser|just a moment|attention required|verify you are (a )?human|press & hold|you have been blocked|access denied|unusual traffic)\b").unwrap()
});

fn decode_entities(s: &str) -> String {
    // Order matters: &amp; last so earlier replacements aren't double-decoded.
    s.replace("&#x27;", "'")
        .replace("&#39;", "'")
        .replace("&quot;", "\"")
        .replace("&nbsp;", " ")
        .replace("&mdash;", "—")
        .replace("&ndash;", "–")
        .replace("&hellip;", "…")
        .replace("&lt;", "<")
        .replace("&gt;", ">")
        .replace("&amp;", "&")
}

/// Convert an HTML document to readable plain text (a small regex pipeline, not
/// a parser).
pub fn html_to_text(html: &str) -> String {
    let mut s = html.to_string();

    for re in DROP_RES.iter() {
        s = re.replace_all(&s, " ").into_owned();
    }
    s = COMMENT_RE.replace_all(&s, " ").into_owned();

    // <a href="X">label</a> → "label (X)".
    s = LINK_RE
        .replace_all(&s, |caps: &Captures| {
            let href = &caps[1];
            let text = TAG_RE.replace_all(&caps[2], "").trim().to_string();
            if text.is_empty() {
                String::new()
            } else if href.starts_with("http") {
                format!("{text} ({href})")
            } else {
                text
            }
        })
        .into_owned();

    s = BLOCK_CLOSE_RE.replace_all(&s, "\n").into_owned();
    s = BR_RE.replace_all(&s, "\n").into_owned();
    s = LI_RE.replace_all(&s, "\n- ").into_owned();

    s = TAG_RE.replace_all(&s, "").into_owned();
    s = decode_entities(&s);

    let collapsed: Vec<String> = s
        .split('\n')
        .map(|line| INLINE_WS_RE.replace_all(line, " ").trim().to_string())
        .collect();
    let joined = collapsed.join("\n");
    BLANKS_RE.replace_all(&joined, "\n\n").trim().to_string()
}

/// Heuristic: an anti-bot / JS-shell placeholder instead of the page.
pub fn looks_blocked(text: &str, is_html: bool) -> bool {
    if is_html && text.chars().count() < THIN_TEXT {
        return true;
    }
    text.chars().count() < 1500 && BLOCKED_RE.is_match(text)
}

/// Validate/normalize the URL: require http(s), upgrade http→https.
fn normalize_url(raw: &str) -> Option<String> {
    let mut u = reqwest::Url::parse(raw.trim()).ok()?;
    match u.scheme() {
        "https" => {}
        "http" => {
            u.set_scheme("https").ok()?;
        }
        _ => return None,
    }
    Some(u.to_string())
}

/// Fetch a URL and return its content as text.
pub async fn execute_web_fetch(client: &reqwest::Client, args: &serde_json::Value) -> ToolResult {
    let Some(url) = normalize_url(args.get("url").and_then(|v| v.as_str()).unwrap_or("")) else {
        return ToolResult::error("Error: a valid http(s) url is required.");
    };

    if let Some((at, content)) = CACHE.lock().unwrap().get(&url) {
        if at.elapsed() < CACHE_TTL {
            return ToolResult::ok(content.clone());
        }
    }

    let resp = match client
        .get(&url)
        .header("User-Agent", UA)
        .header("Accept", "text/html,*/*")
        .timeout(REQUEST_TIMEOUT)
        .send()
        .await
    {
        Ok(r) => r,
        Err(_) => {
            return ToolResult::error(format!(
                "Error: failed to fetch {url} (network error or timeout)."
            ))
        }
    };

    if !resp.status().is_success() {
        let status = resp.status();
        let reason = status.canonical_reason().unwrap_or("");
        return ToolResult::error(format!(
            "Error: {url} returned HTTP {} {reason}.",
            status.as_u16()
        ));
    }

    let ctype = resp
        .headers()
        .get("content-type")
        .and_then(|v| v.to_str().ok())
        .unwrap_or("")
        .to_string();
    let raw = resp.text().await.unwrap_or_default();
    let is_html = ctype.to_lowercase().contains("html");
    let body = if is_html {
        html_to_text(&raw)
    } else {
        raw.trim().to_string()
    };

    if body.is_empty() {
        return ToolResult::ok(format!(
            "Fetched {url} but it had no extractable text content."
        ));
    }
    if looks_blocked(&body, is_html) {
        return ToolResult::error(format!(
            "Error: {url} requires JavaScript or blocks automated access — its content cannot be fetched. Do not retry this URL or other pages on this site; use a different source instead."
        ));
    }

    let char_count = body.chars().count();
    let truncated = if char_count > MAX_CHARS {
        let head: String = body.chars().take(MAX_CHARS).collect();
        format!(
            "{head}\n\n[truncated — {} more chars]",
            char_count - MAX_CHARS
        )
    } else {
        body
    };
    let content = format!("Content of {url}:\n\n{truncated}");
    CACHE
        .lock()
        .unwrap()
        .insert(url, (Instant::now(), content.clone()));
    ToolResult::ok(content)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn html_to_text_pipeline() {
        let html = r#"<head><title>T</title></head><body>
            <script>var x = 1;</script>
            <p>Hello &amp; welcome</p>
            <a href="https://example.com">click here</a>
            <ul><li>one</li><li>two</li></ul>
            <div>line</div><br>after
        </body>"#;
        let text = html_to_text(html);
        assert!(!text.contains("var x"), "script must be dropped");
        assert!(
            text.contains("Hello & welcome"),
            "entities decoded, p → line"
        );
        assert!(
            text.contains("click here (https://example.com)"),
            "link formatted"
        );
        assert!(text.contains("- one"));
        assert!(text.contains("- two"));
        assert!(text.contains("line"));
        assert!(text.contains("after"));
    }

    #[test]
    fn relative_link_keeps_only_text() {
        let text = html_to_text(r#"<a href="/local">label</a>"#);
        assert_eq!(text, "label");
    }

    #[test]
    fn decode_entities_order() {
        assert_eq!(decode_entities("a &amp; b &lt;c&gt;"), "a & b <c>");
        assert_eq!(decode_entities("&#39;q&quot;"), "'q\"");
    }

    #[test]
    fn looks_blocked_thin_html_and_botwall() {
        assert!(looks_blocked("tiny", true)); // thin HTML
        assert!(!looks_blocked("tiny", false)); // short non-HTML is fine
        assert!(looks_blocked(
            "Just a moment... checking your browser",
            false
        ));
        assert!(!looks_blocked(&"x".repeat(2000), true)); // long real page
    }

    #[test]
    fn normalize_url_upgrades_and_validates() {
        assert!(normalize_url("http://example.com")
            .unwrap()
            .starts_with("https://"));
        assert!(normalize_url("https://example.com/path").is_some());
        assert!(normalize_url("ftp://example.com").is_none());
        assert!(normalize_url("not a url").is_none());
    }
}
